File: app/src/main/java/no/joharei/flixr/photosets/PhotosetsPresenter.java
package no.joharei.flixr.photosets;

import android.util.Log;

import javax.inject.Inject;

import no.joharei.flixr.MainApplication;
import no.joharei.flixr.photos.PhotosPresenter;
import no.joharei.flixr.tools.RxAssist;

public class PhotosetsPresenter {
    private static final String TAG = PhotosPresenter.class.getSimpleName();
    @Inject
    PhotosetsApi mainApi;
    private PhotosetsView view;

    void attachView(PhotosetsView view) {
        this.view = view;
        MainApplication.component(view.getContext()).inject(this);
    }

    void fetchPhotosets(String userId) {
        mainApi.getPhotosets(userId)
                .compose(RxAssist.applyDefaultSchedulers())
                .subscribe(photosets -> {
                    view.showPhotosets(photosets.getPhotosets());
                }, throwable -> {
                    Log.e(TAG, "Failed fetching photosets", throwable);
                    // TODO
                });
    }
}

File: app/src/main/java/no/joharei/flixr/photosets/PhotosetsPresenter.kt
package no.joharei.flixr.photosets

import io.reactivex.disposables.CompositeDisposable
import no.joharei.flixr.MainApplication
import no.joharei.flixr.tools.applyDefaultSchedulers
import timber.log.Timber
import javax.inject.Inject

class PhotosetsPresenter {
    @Inject
    internal lateinit var photosetsApi: PhotosetsApi
    private lateinit var view: PhotosetsView
    private val compositeSubscription = CompositeDisposable()

    internal fun attachView(view: PhotosetsView) {
        this.view = view
        MainApplication.component.inject(this)
    }

    internal fun fetchPhotosets(userId: String?) {
        view.showProgress()
        compositeSubscription.add(photosetsApi.getPhotosets(userId)
                .applyDefaultSchedulers()
                .subscribe(
                        { photosets ->
                            view.hideProgress()
                            view.showPhotosets(photosets.photosets)
                        },
                        { throwable ->
                            Timber.e(throwable, "Failed fetching photosets")
                            photosetsApi.clearCache()
                            // TODO
                        }))
    }

    internal fun stop() {
        compositeSubscription.clear()
    }
}
